src/persist-client/src/internal/state_fetcher.rs

```diff
@@ -0,0 +1,373 @@
+        *state = self.fetch_latest_state(&state.shard_id, hint).await?;
+        Ok(())
+    }
+
+    pub fn apply_diffs<'a, I: IntoIterator<Item = &'a VersionedData>>(
+        &self,
+        state: &mut State<K, V, T, D>,
+        diffs: I,
+    ) {
+        let mut state_seqno = state.seqno;
+        let diffs = diffs.into_iter().filter_map(move |diff| {
+            if diff.seqno == state_seqno {
+                // No-op.
+                return None;
+            }
+            let diff = self.decode_diff(diff);
+            assert_eq!(diff.seqno_from, state_seqno);
+            state_seqno = diff.seqno_to;
+            Some(diff)
+        });
+        state
+            .apply_diffs(&self.metrics, diffs)
+            .expect("state diff should apply cleanly");
+    }
+
+    // WIP Returns all diffs and a rollup at the version == the first diff
+    pub async fn fetch_live_states(
+        &self,
+        shard_id: &ShardId,
+    ) -> Result<(State<K, V, T, D>, Vec<VersionedData>), CodecMismatch> {
+        let retry = self
+            .metrics
+            .retries
+            .fetch_live_states
+            .stream(Retry::persist_defaults(SystemTime::now()).into_retry_stream());
+        loop {
+            let live_diffs = self.fetch_live_diffs(&shard_id).await;
+            let earliest_live_diff = match live_diffs.first() {
+                Some(x) => x,
+                None => {
+                    // Live diffs is only empty if the shard is uninitialized.
+                    // Initialize it and try again. This throws away the return
+                    // value of maybe_init, but that's fine because we should
+                    // never hit this case in practice.
+                    let _ = self.maybe_init(shard_id).await?;
+                    // Intentionally don't sleep on retry.
+                    retry.retries.inc();
+                    continue;
+                }
+            };
+            let state = match self
+                .fetch_rollup_at_seqno(shard_id, live_diffs.clone(), earliest_live_diff.seqno)
+                .await
+            {
+                Some(x) => x?,
+                None => {
+                    // We maintain an invariant that a rollup always exists for
+                    // the earliest live diff. Since we didn't find out, that
+                    // can only mean that the live_diffs we just fetched are
+                    // obsolete (there's a race condition with gc). This should
+                    // be rare in practice, so inc a counter and try again.
+                    // Intentionally don't sleep on retry.
+                    retry.retries.inc();
+                    continue;
+                }
+            };
+            assert_eq!(earliest_live_diff.seqno, state.seqno);
+            return Ok((state, live_diffs));
+        }
+    }
+
+    // WIP only callable on initialized shards, so panics if hint is empty
+    async fn fetch_latest_state(
```

Comment:
so tbh I've spent quite a lot of time trying to understand StateFetcher and have gotten very turned about a few times, and I think some of it is just naming. I find this one tricky because it doesn't fetch the _latest_ state, it constructs the latest state up through `hint`, and what hint is isn't obvious from the type.

this might be easier to parse if it were something like:

```
async fn state(&self, shard_id, diffs: Vec<StateDiff>)
```

Reply:
I've tried to solve this one with rustdoc, too. grok-ability matters a great deal to me, but to be entirely honest, I've probably passed 32 hours of tweaking state_fetcher/state_versions at this point, and run out of motivation for more tweaks. i'd love to discuss a concrete proposal (i.e. a diff that passes CI) if you have one <3